mesonbuild/interpreterbase/interpreterbase.py

```diff
@@ -123,7 +123,19 @@ def sanity_check_ast(self) -> None:
             raise InvalidCode('No statements in code.')
         first = self.ast.lines[0]
         if not isinstance(first, mparser.FunctionNode) or first.func_name != 'project':
-            raise InvalidCode('First statement must be a call to project')
+            p = pathlib.Path(self.source_root).resolve()
+            found = p
+            for parent in p.parents:
+                if (parent / 'meson.build').is_file():
+                    found = parent
+                else:
+                    break
+
+            error = 'first statement must be a call to project()'
+            if found != p:
+                raise InvalidCode(f'Not the project root: {error}\n\nDid you mean to run meson from the directory: "{found}"?')
```

Comment:
We should not recommend an alternate path to run meson from without first checking whether that file also has a `project()` in it, otherwise it will just confuse users.

Even if that file has `project()` in it, it may not be the project that the user wanted to build.

Reply:
Also, the message implies that you must run meson from the project's root directory, which is not true. Some projects have the meson build files in a subdir of the project root (such as VLC).

Reply:
I'd be open to checking that one for a project() function, not sure of the best way to do it though.

Since meson doesn't tell you to build that one, merely asks you if you "meant" to, I don't believe it's a concern that it might be the wrong but valid project.

Reply:
If we're doing heuristics, just opening the file and checking that the first line starts with `project(` might be enough.

We expect users to be "smart enough" but if you tell them that they can try something, they will try it, and then get even more confused when it doesn't work. Sometimes more information is worse. If they're not smart enough to figure out the topmost meson.build file themselves, giving them the wrong location will make things worse.

Reply:
Makes sense. That would be a nice and simple improvement and continue to catch the vast majority of cases.